Dead interpolation elimination improvements

beccadax opened the pull request:
The optimizer doesn't understand the effects of `appendLiteral(_:)` and `appendInterpolation(_:)` calls, which only modify `self`. Because of this, it can't tell that an interpolation can be eliminated if the resulting string is unused, which causes a (no-assertions-only) test failure and some code size regressions.

This is an attempt to help DeadCodeElimination along. The first revision benchmarked poorly when I tried it a few days ago; I'm just pushing it now to get a baseline, and then I'll experiment a little more from there. Currently pretty hacky, too.

This could be a dead end. If it isn't, it should resolve [SR-9008](https://bugs.swift.org/browse/SR-9008), rdar://45294605, and rdar://45788348.

beccadax committed:
WIP: Dead interpolation elimination improvements

brentdax commented:
Wow, much better than expected. Let's do a longer run.

brentdax commented:
@swift-ci please benchmark

beccadax committed:
Rename new semantic to interpolation.append

@eeckstein pointed out that it’s not strictly correct to declare a guarantee that a function wrapping a call to write(to:) on an arbitrary type has no side effects except on self. However, we do currently eliminate these calls anyway and we want to keep doing so, so this is best thought of as a quirk of interpolation behavior, not a result of the effects system.

Result: We’re going to stick with using an @_semantics attribute, but rename it to avoid talking about effects.

beccadax committed:
Re-enable dead interpolation test

beccadax committed:
Add new elimination test for non-small string

If an interpolation is estimated to contain more than 15 characters, we will generate code which is harder to correctly eliminate. Make sure we handle this case.

beccadax committed:
Make DefaultStringInterpolation.init easier to eliminate

The optimizer has a hard time understanding that _StringGuts.init(_initialCapacity:) is fully eliminatable, particularly when we end up choosing the large string path. This change allows us to eliminate the interpolation before we do that inlining. With it, even large string allocations can be quickly and completely eliminated.

beccadax committed:
[SILOptimizer] Clean up interpolation code

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please benchmark

brentdax commented:
I'm reasonably satisfied with these results. Comments on some surprises: 

* **Smaller-than-expected `libswiftSwiftReflectionTest.dylib` improvement**: The baseline changed. On Sunday, master was building a 53k dylib; on Thursday, it built a 49k dylib. In both runs, this PR built a 45k version.
* **`DictionaryKeysContains.o` size regression**: The inliner seems to make different decisions about `Dictionary.init<A>(uniqueKeysWithValues:)`—in the branch, it doesn't inline `_NativeDictionary.merge<A>(_:isUnique:uniquingKeysWith:)` but instead keeps it as a separate specialized function. There's a string interpolation in the `merge` method which generates a `fatalError()` message, and it looks like it's not inlining calls to `DefaultStringInterpolation.init`, so perhaps that's changing the inlining cost enough so that it makes a different decision? (In any case, I noticed some slightly questionable code in some of these members, so it might be changing soon...)

brentdax commented:
The WordCount.o, SequenceAlgos.o, and TwoSum.o tests are heavy on calls to `CheckResults(_:)`, a function with a string interpolation in its (never-taken) failure branch. These are all sensitive to changes in how much interpolation code we inline.

brentdax commented:
@swift-ci please benchmark

beccadax committed:
[SILOptimizer] Make interpolation detection cleaner

brentdax commented:
@swift-ci please test

brentdax commented:
That's the CI's usual "bad hash" failure. Tests are running [here](https://ci.swift.org/job/swift-PR-osx/8921/console).

eeckstein commented on lib/SILOptimizer/Transforms/DeadObjectElimination.cpp:
```diff
@@ -224,6 +269,28 @@ static bool canZapInstruction(SILInstruction *Inst, bool acceptRefCountInsts) {
+      return true;
+
+  // We can remove loads of addresses within the allocation.
+  if (auto *Load = dyn_cast<LoadInst>(Inst))
```
I'm wondering why we can't remove load instructions in general, i.e. remove the "!Inst->mayReadFromMemory()" from the condition above.

eeckstein commented on lib/SILOptimizer/Transforms/DeadObjectElimination.cpp:
```diff
@@ -224,6 +269,28 @@ static bool canZapInstruction(SILInstruction *Inst, bool acceptRefCountInsts) {
+            // memory management. We don't do that here (that's what
+            // DeadObjectAnalysis below is supposed to do), but we know we don't
+            // need to for DefaultStringInterpolation, because we know that the
+            // initialization of its sole property will be eliminated too.
```
I don't understand this comment. Can you give an example?

brentdax replied:
I may be a little confused here. What I'm trying to say is that in the general case, if we eliminate a load and a release, we should look for corresponding stores and replace them with releases of the values that would have been stored into them. We don't need to do that for `DefaultStringInterpolation` because we know that if we eliminate its allocation, we will also eliminate anything that could have allocated or retained the string storage object that would have been inside it, because that only happens inside eliminatable calls to `appendLiteral(_:)`, `appendInterpolation(_:)`, or `init(literalCapacity:interpolationCount:)`.

eeckstein replied:
I see. Can you change the comment to this explanation?

eeckstein commented:
> DictionaryKeysContains.o

I think we can live with the code size regression for now.

beccadax committed:
wip

Making nontrivial dead object elimination work.

beccadax committed:
WIP: Switch to nontrivial dead object elim

Rather than trying to hack trivial dead object elimination to handle certain nontrivial objects, I’m now trying to use the nontrivial dead object elimination tools. These ought to be adequate for our use case; the comments indicate that their main weakness is destructor analysis, which I don’t think should be a factor for structs, and certainly isn’t for interpolations (which always use native strings). So far this passes all interpolation tests, but breaks when applied to arbitrary structs, so I don’t fully trust it yet.

brentdax commented:
@swift-ci please benchmark

brentdax commented:
What CI meant to say was: 

# Summary for master full

**Unexpected test results, excluded stats for Tagged, NonEmpty, ProcedureKitCloud, GRDB, ModelAssistant, Wordy**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-c2a054c1c6ee9d1)
    - [debug-batch brief](#debug-batch-brief-c2a054c1c6ee9d1)
    - [debug-batch detailed](#debug-batch-detailed-c2a054c1c6ee9d1)
- [Release](#release-c2a054c1c6ee9d1)
    - [release brief](#release-brief-c2a054c1c6ee9d1)
    - [release detailed](#release-detailed-c2a054c1c6ee9d1)



<a name="debug-batch-c2a054c1c6ee9d1">

# Debug-batch

<a name="debug-batch-brief-c2a054c1c6ee9d1">

## debug-batch brief
<details open>
<summary>Regressed (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 21,299,318,627,006 | 23,930,268,593,075 | 2,630,949,966,069 | 12.35% :no_entry:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
LLVM.NumLLVMBytesOutput | 911,949,140 | 911,727,304 | -221,836 | -0.02%
time.swift-driver.wall | 2246.9s | 2256.4s | 9.5s | 0.42%
</details>

<a name="debug-batch-detailed-c2a054c1c6ee9d1">

## debug-batch detailed
<details>
<summary>Regressed (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 21,299,318,627,006 | 23,930,268,593,075 | 2,630,949,966,069 | 12.35% :no_entry:
</details>
<details>
<summary>Improved (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Driver.NumDriverPipePolls | 304,783 | 292,005 | -12,778 | -4.19% :white_check_mark:
Driver.NumDriverPipeReads | 345,376 | 331,797 | -13,579 | -3.93% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (92)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 45,359,895,288 | 45,451,505,008 | 91,609,720 | 0.2%
AST.NumDecls | 65,849 | 65,849 | 0 | 0.0%
AST.NumDependencies | 152,683 | 152,689 | 6 | 0.0%
AST.NumImportedExternalDefinitions | 960,730 | 960,730 | 0 | 0.0%
AST.NumInfixOperators | 24,640 | 24,640 | 0 | 0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 180,845 | 180,845 | 0 | 0.0%
AST.NumLocalTypeDecls | 113 | 113 | 0 | 0.0%
AST.NumObjCMethods | 12,528 | 12,528 | 0 | 0.0%
AST.NumPostfixOperators | 13 | 13 | 0 | 0.0%
AST.NumPrecedenceGroups | 12,687 | 12,687 | 0 | 0.0%
AST.NumPrefixOperators | 71 | 71 | 0 | 0.0%
AST.NumReferencedDynamicNames | 101 | 101 | 0 | 0.0%
AST.NumReferencedMemberNames | 3,046,567 | 3,046,567 | 0 | 0.0%
AST.NumReferencedTopLevelNames | 210,981 | 210,981 | 0 | 0.0%
AST.NumSourceBuffers | 288,503 | 288,503 | 0 | 0.0%
AST.NumSourceLines | 2,107,547 | 2,107,547 | 0 | 0.0%
AST.NumSourceLinesPerSecond | 1,167,653 | 1,160,805 | -6,848 | -0.59%
AST.NumTotalClangImportedEntities | 3,562,456 | 3,565,535 | 3,079 | 0.09%
AST.NumUsedConformances | 185,274 | 185,274 | 0 | 0.0%
Driver.ChildrenMaxRSS | 68,565,780,480 | 68,615,297,024 | 49,516,544 | 0.07%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 13,529 | 13,529 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 352,809,515,144 | 353,114,816,472 | 305,301,328 | 0.09%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 92,680 | 92,680 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 3,587,318 | 3,587,289 | -29 | -0.0%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 1,632,326 | 1,632,326 | 0 | 0.0%
IRModule.NumIRGlobals | 1,857,354 | 1,857,354 | 0 | 0.0%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 42,089,475 | 42,077,011 | -12,464 | -0.03%
IRModule.NumIRNamedMetaData | 66,177 | 66,177 | 0 | 0.0%
IRModule.NumIRValueSymbols | 3,120,194 | 3,120,194 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 911,949,140 | 911,727,304 | -221,836 | -0.02%
Parse.NumFunctionsParsed | 2,170,339 | 2,170,339 | 0 | 0.0%
Parse.NumIterableDeclContextParsed | 865,730 | 865,730 | 0 | 0.0%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 1,286,682 | 1,286,682 | 0 | 0.0%
SILModule.NumSILGenGlobalVariables | 26,821 | 26,821 | 0 | 0.0%
SILModule.NumSILGenVtables | 10,171 | 10,171 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 37,360 | 37,360 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 1,170,065 | 1,171,362 | 1,297 | 0.11%
SILModule.NumSILOptGlobalVariables | 27,550 | 27,550 | 0 | 0.0%
SILModule.NumSILOptVtables | 16,355 | 16,355 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 72,507 | 72,507 | 0 | 0.0%
Sema.AccessLevelRequest | 1,912,680 | 1,913,194 | 514 | 0.03%
Sema.DefaultAndMaxAccessLevelRequest | 45,840 | 45,843 | 3 | 0.01%
Sema.EnumRawTypeRequest | 13,205 | 13,205 | 0 | 0.0%
Sema.ExtendedNominalRequest | 2,675,157 | 2,675,431 | 274 | 0.01%
Sema.InheritedDeclsReferencedRequest | 84,613,324 | 84,668,705 | 55,381 | 0.07%
Sema.InheritedTypeRequest | 441,224 | 441,287 | 63 | 0.01%
Sema.IsDynamicRequest | 1,535,733 | 1,535,733 | 0 | 0.0%
Sema.IsObjCRequest | 1,333,835 | 1,333,978 | 143 | 0.01%
Sema.NamedLazyMemberLoadFailureCount | 18,161 | 18,184 | 23 | 0.13%
Sema.NamedLazyMemberLoadSuccessCount | 13,018,294 | 13,012,199 | -6,095 | -0.05%
Sema.NominalTypeLookupDirectCount | 24,981,204 | 24,991,188 | 9,984 | 0.04%
Sema.NumConformancesDeserialized | 3,879,389 | 3,882,556 | 3,167 | 0.08%
Sema.NumConstraintScopes | 13,045,831 | 13,046,856 | 1,025 | 0.01%
Sema.NumConstraintsConsideredForEdgeContraction | 29,316,344 | 29,316,587 | 243 | 0.0%
Sema.NumDeclsDeserialized | 30,827,826 | 30,873,244 | 45,418 | 0.15%
Sema.NumDeclsValidated | 1,645,646 | 1,645,641 | -5 | -0.0%
Sema.NumFunctionsTypechecked | 920,531 | 920,531 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 872,637 | 872,284 | -353 | -0.04%
Sema.NumLazyGenericEnvironments | 6,157,722 | 6,157,904 | 182 | 0.0%
Sema.NumLazyGenericEnvironmentsLoaded | 165,400 | 165,424 | 24 | 0.01%
Sema.NumLazyIterableDeclContexts | 4,932,562 | 4,933,380 | 818 | 0.02%
Sema.NumLeafScopes | 8,944,710 | 8,945,570 | 860 | 0.01%
Sema.NumTypesDeserialized | 10,882,131 | 10,888,374 | 6,243 | 0.06%
Sema.NumTypesValidated | 1,078,137 | 1,078,132 | -5 | -0.0%
Sema.NumUnloadedLazyIterableDeclContexts | 3,500,966 | 3,499,675 | -1,291 | -0.04%
Sema.OverriddenDeclsRequest | 4,103,315 | 4,121,991 | 18,676 | 0.46%
Sema.RequirementRequest | 54,623 | 54,619 | -4 | -0.01%
Sema.SelfBoundsFromWhereClauseRequest | 50,408,559 | 50,467,490 | 58,931 | 0.12%
Sema.SetterAccessLevelRequest | 108,626 | 108,626 | 0 | 0.0%
Sema.SuperclassDeclRequest | 66,573,449 | 66,597,424 | 23,975 | 0.04%
Sema.SuperclassTypeRequest | 30,277 | 30,277 | 0 | 0.0%
Sema.TypeDeclsFromWhereClauseRequest | 26,254 | 26,257 | 3 | 0.01%
Sema.USRGenerationRequest | 6,141,318 | 6,181,319 | 40,001 | 0.65%
Sema.UnderlyingTypeDeclsReferencedRequest | 2,296,019 | 2,296,093 | 74 | 0.0%
</details>

<a name="release-c2a054c1c6ee9d1">

# Release

<a name="release-brief-c2a054c1c6ee9d1">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 24,042,045,391,202 | 24,042,730,935,196 | 685,543,994 | 0.0%
LLVM.NumLLVMBytesOutput | 811,141,394 | 811,010,570 | -130,824 | -0.02%
time.swift-driver.wall | 4466.2s | 4465.0s | -1.2s | -0.03%
</details>

<a name="release-detailed-c2a054c1c6ee9d1">

## release detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (23)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumImportedExternalDefinitions | 178,485 | 178,485 | 0 | 0.0%
AST.NumLoadedModules | 11,404 | 11,404 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 602,601 | 602,601 | 0 | 0.0%
AST.NumUsedConformances | 186,004 | 186,004 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 2,985,101 | 2,985,023 | -78 | -0.0%
IRModule.NumIRFunctions | 1,338,631 | 1,338,723 | 92 | 0.01%
IRModule.NumIRGlobals | 1,465,764 | 1,465,655 | -109 | -0.01%
IRModule.NumIRInsts | 28,671,707 | 28,672,463 | 756 | 0.0%
IRModule.NumIRValueSymbols | 2,620,363 | 2,620,358 | -5 | -0.0%
LLVM.NumLLVMBytesOutput | 811,141,394 | 811,010,570 | -130,824 | -0.02%
SILModule.NumSILGenFunctions | 561,090 | 561,090 | 0 | 0.0%
SILModule.NumSILOptFunctions | 695,820 | 696,113 | 293 | 0.04%
Sema.NumConformancesDeserialized | 1,574,881 | 1,572,972 | -1,909 | -0.12%
Sema.NumConstraintScopes | 11,500,306 | 11,500,306 | 0 | 0.0%
Sema.NumDeclsDeserialized | 3,928,331 | 3,928,720 | 389 | 0.01%
Sema.NumDeclsValidated | 855,460 | 855,460 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 450,388 | 450,388 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 144,163 | 144,163 | 0 | 0.0%
Sema.NumLazyGenericEnvironments | 792,879 | 792,879 | 0 | 0.0%
Sema.NumLazyGenericEnvironmentsLoaded | 15,999 | 15,999 | 0 | 0.0%
Sema.NumLazyIterableDeclContexts | 531,886 | 531,906 | 20 | 0.0%
Sema.NumTypesDeserialized | 2,136,447 | 2,139,197 | 2,750 | 0.13%
Sema.NumTypesValidated | 401,820 | 401,820 | 0 | 0.0%
</details>

brentdax commented:
Also, ModelAssistant UPASSed. So, that happened.

brentdax commented:
So, this passes all tests and doesn't seem to cause any problems as long as `NONTRIVIAL_ALLOC_STACK_FOR_EVERYONE` is not defined. But *with* it defined, it seems to introduce incorrect releases in some tests, which makes me worried there's an underlying issue. Reduction for one of these (pop it into test/stdlib and run it with lit; make sure you have a release-mode, no-assertion stdlib):

```swift
// RUN: %target-run-simple-swift
// REQUIRES: executable_test

// REQUIRES: objc_interop

// FIXME: rdar://problem/31311598
// UNSUPPORTED: OS=ios
// UNSUPPORTED: OS=tvos
// UNSUPPORTED: OS=watchos

//
// Tests for the NSString APIs as exposed by String
//

import Foundation

_ = "абвг" == "абВГ".lowercased(with: Locale.current)
```

As far as I can tell, the failing tests all involve bridged NSStrings, and interpolation buffers should ~never be NSStrings. But it's still a concern, and this patch is probably getting too large to land in swift-5.0-branch anyway.
